drivers/adc/adc_demo/adc_demo.c

```diff
@@ -85,6 +112,129 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
 	return SUCCESS;
 }
 
+int32_t adc_get_number_of_channels(void* dev)
+{
+	struct adc_demo_desc *desc = dev;
+
+	return desc->active_ch;
+}
+
+/***************************************************************************//**
+ * @brief close all channels
+ * @param dev - physical instance of an adc device
+ * @return SUCCESS in case of success.
+*******************************************************************************/
+int32_t close_adc_channels(void* dev)
+{
+	struct adc_demo_desc *desc = dev;
+	desc->active_ch = 0;
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * @brief utility function for computing next upcoming channel
+ * @param ch_mask - active channels .
+ * @param last_idx -  previous index.
+ * @param new_idx - upcoming channel index, return param.
+ * @return 1 if there are more channels, 0 if done.
+*******************************************************************************/
+static bool get_next_ch_idx(uint32_t ch_mask, uint32_t last_idx,
+			    uint32_t *new_idx)
+{
+	last_idx++;
+	ch_mask >>= last_idx;
+	if (!ch_mask) {
+		*new_idx = -1;
+		return 0;
+	}
+	while (!(ch_mask & 1)) {
+		last_idx++;
+		ch_mask >>= 1;
+	}
+	*new_idx = last_idx;
+
+	return 1;
+}
+
+/***************************************************************************//**
+ * @brief function for reading samples
+ * @param dev - physical instance of adc device
+ * @param buff - buffer for reading samples
+ * @param samples - number of samples to receive
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adc_read_samples(void* dev, uint16_t* buff, uint32_t samples)
```

Comment:
Maybe return an error code instead or `samples`.

Otherwise, you should switch the return type for the function to `uint32_t` since the number of samples cannot be negative.

Also, the header comment @return does not match the actual implementation.